src/System.Private.CoreLib/src/Internal/Runtime/Augments/RuntimeThread.cs

```diff
@@ -0,0 +1,127 @@
+
+using System;
+using System.Diagnostics;
+using System.Runtime;
+using System.Threading;
+using Microsoft.Win32.SafeHandles;
+
+namespace Internal.Runtime.Augments
+{
+    public sealed partial class RuntimeThread
+    {
+        [ThreadStatic]
+        private static RuntimeThread t_currentThread;
+
+        /// <summary>
+        /// Used by <see cref="WaitHandle"/>'s multi-wait functions
+        /// </summary>
+        private WaitHandleArray<SafeWaitHandle> _waitedSafeWaitHandles;
+
+        private ThreadState _threadState;
+
+        private RuntimeThread()
+        {
+            _waitedSafeWaitHandles = new WaitHandleArray<SafeWaitHandle>(elementInitializer: null);
+            _threadState = ThreadState.Unstarted;
+
+#if PLATFORM_UNIX
+            _waitInfo = new WaitSubsystem.ThreadWaitInfo(this);
+#else
+            _waitedHandles = new WaitHandleArray<IntPtr>(elementInitializer: null);
+#endif
+        }
+
+        /// <summary>
+        /// Callers must ensure to clear the array after use
+        /// </summary>
+        internal SafeWaitHandle[] GetWaitedSafeWaitHandleArray(int requiredCapacity)
+        {
+            Debug.Assert(this == CurrentThread);
+
+            _waitedSafeWaitHandles.VerifyElementsAreDefault();
+            _waitedSafeWaitHandles.EnsureCapacity(requiredCapacity);
+            return _waitedSafeWaitHandles.Items;
+        }
+
+        public static RuntimeThread Create(ThreadStart start) { throw null; }
+        public static RuntimeThread Create(ThreadStart start, int maxStackSize) { throw null; }
+        public static RuntimeThread Create(ParameterizedThreadStart start) { throw null; }
+        public static RuntimeThread Create(ParameterizedThreadStart start, int maxStackSize) { throw null; }
+
+        public static RuntimeThread CurrentThread
+        {
+            get
+            {
+                RuntimeThread currentThread = t_currentThread;
+                if (currentThread != null)
+                {
+                    return currentThread;
+                }
+
+                t_currentThread = currentThread = new RuntimeThread();
+                Debug.Assert(currentThread._threadState == ThreadState.Unstarted);
+                currentThread._threadState = ThreadState.Running | ThreadState.Background;
+                return currentThread;
+            }
+        }
+
+        public bool IsAlive { get { throw null; } }
+        public bool IsBackground { get { throw null; } set { throw null; } }
+        public bool IsThreadPoolThread { get { throw null; } }
+        public int ManagedThreadId { get { throw null; } }
+        public string Name { get { throw null; } set { throw null; } }
+        public ThreadPriority Priority { get { throw null; } set { throw null; } }
+
+        public ThreadState ThreadState
+        {
+            get
+            {
+                Interlocked.MemoryBarrier(); // read the latest value
```

Comment:
All you should need here is to ensure that the codegen will not optimize fetch of _threadState out, so that patterns like active waiting on it works fine. Volatile.Read is sufficient to do that. I do not see why you need a MemoryBarrier.